python/python/bystro/ancestry/inference.py

```diff
@@ -107,35 +104,47 @@ def _package_ancestry_response_from_pop_probs(
             }
         )
         ancestry_results.append(
-            AncestryResult(
+            AncestryScoresOneSample(
                 sample_id=sample_id,
-                top_hit=AncestryTopHit(
-                    probability=max_value,
-                    populations=top_pops
-                ),
+                top_hit=AncestryTopHit(probability=max_value, populations=top_pops),
                 populations=pop_vector,
                 superpops=superpop_vector,
                 missingness=missingnesses[sample_id],
             )
         )
 
-    return AncestryResponse(vcf_path=str(vcf_path), results=ancestry_results, pcs=pcs_for_plotting)
+    return AncestryResults(results=ancestry_results, pcs=pcs_for_plotting)
 
 
-# TODO: implement with ray
-def infer_ancestry(
-    ancestry_model: AncestryModel, genotypes: pd.DataFrame, vcf_path: Path | str
-) -> AncestryResponse:
+def infer_ancestry(ancestry_model: AncestryModel, genotypes: Dataset) -> AncestryResults:
     """Run an ancestry job."""
-    # TODO: main ancestry model logic goes here.  Just stubbing out for now.
 
-    logger.debug("Filling missing data for VCF: %s", vcf_path)
+    logger.debug("Beginning ancestry inference")
     with Timer() as timer:
-        imputed_genotypes, missingnesses = _fill_missing_data(genotypes)
-    logger.debug("Finished filling missing data for VCF in %f seconds", timer.elapsed_time)
-    pcs_for_plotting, pop_probs_df = ancestry_model.predict_proba(imputed_genotypes)
+        mask = pc.field("locus").isin(ancestry_model.pca_loadings_df.index)
+        genotypes = genotypes.filter(mask).to_table().to_pandas()
+        genotypes = genotypes.set_index("locus")
+        # TODO: @akotlar 2024-01-31: Replace reliance on imputation with Austin Talbot's model
+        # which is robust to missing data.
+        missing_rows = list(set(ancestry_model.pca_loadings_df.index) - set(genotypes.index))
+        if missing_rows:
+            missing_rows_df = pd.DataFrame(
+                np.ones((len(missing_rows), len(genotypes.columns))),
+                index=missing_rows,
+                columns=genotypes.columns,
+            )
+            genotypes = pd.concat([genotypes, missing_rows_df])
+
+        missingness = len(missing_rows) / len(ancestry_model.pca_loadings_df.index)
+        sample_missingnesses = genotypes.isna().mean(axis=0) + missingness
+
+        genotypes[genotypes.isna()] = 1
+
+        pcs_for_plotting, pop_probs_df = ancestry_model.predict_proba(genotypes)
+    logger.debug("Completed ancestry inference in %f seconds", timer.elapsed_time)
+
     return _package_ancestry_response_from_pop_probs(
-        vcf_path, pcs_for_plotting, pop_probs_df, missingnesses
+        pcs_for_plotting, pop_probs_df, sample_missingnesses
```

Comment:
While I do prefer sample_missingness over missingness as the term used here, I think it would be helpful to include somewhere that this sample_missingness is relative to the trained model and not the measure of sample quality in any way, maybe not anywhere in the code necessarily but just wanted to mark that we should at least state this in the tutorial/info about the ancestry model somewhere.